algorithms/src/Kokkos_Random.hpp

```diff
@@ -1543,13 +1543,23 @@ template <class ViewType, class RandomPool, class IndexType = int64_t>
 void fill_random(ViewType a, RandomPool g,
                  typename ViewType::const_value_type begin,
                  typename ViewType::const_value_type end) {
-  fill_random(typename ViewType::execution_space{}, a, g, begin, end);
+  Kokkos::fence(
+      "fill_random: fence before since no execution space instance provided");
+  typename ViewType::execution_space exec;
+  fill_random(exec, a, g, begin, end);
+  exec.fence(
+      "fill_random: fence after since no execution space instance provided");
```

Comment:
Once again I'd like to propose a wrapper for this kind of change.

```c++
template <typename execution_space>
class Fencer
{
    public:
        Fencer(std::string msg_) : msg(std::move(msg_))
        {
            Kokkos::fence(msg + ": fence before since no execution space instance provided");
        }

        ~Fencer()
        {
            space.fence(msg + ": fence after since no execution space instance provided");
        }
    public:
        execution_space space {};
    private:
        std::string msg;
};

template <typename execution_space>
void my_function(const execution_space&, std::ostream& out)
{
    out << __PRETTY_FUNCTION__ << std::endl;
}

void my_function(std::ostream& out)
{
    out << __PRETTY_FUNCTION__ << ": enter" << std::endl;
    my_function(Fencer<execution_space>("my_function").space, out);
    out << __PRETTY_FUNCTION__ << ": exit" << std::endl;
}

TEST(Kokkos, fencer)
{
    std::cout << "BEGIN OF TEST" << std::endl;
    my_function(std::cout);
    std::cout << "END OF TEST" << std::endl;
}
```

that would give

```bash
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[ RUN      ] Kokkos.fencer
BEGIN OF TEST
void HELM::tests::utils::my_function(std::ostream&): enter
KokkosP: Executing fence on device 16777216 with unique execution identifier 0
KokkosP:     my_function: fence before since no execution space instance provided
KokkosP: Execution of fence 0 is completed.
KokkosP: Executing fence on device 0 with unique execution identifier 1
KokkosP:     my_function: fence before since no execution space instance provided
KokkosP: Execution of fence 1 is completed.
void HELM::tests::utils::my_function(const execution_space&, std::ostream&) [with execution_space = Kokkos::OpenMP; std::ostream = std::basic_ostream<char>]
KokkosP: Executing fence on device 16777217 with unique execution identifier 2
KokkosP:     my_function: fence after since no execution space instance provided
KokkosP: Execution of fence 2 is completed.
void HELM::tests::utils::my_function(std::ostream&): exit
END OF TEST
[       OK ] Kokkos.fencer (0 ms)
[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (0 ms total)
[  PASSED  ] 1 test.
```

Reply:
@dalg24 @masterleinad @maartenarnst

Reply:
Note that
```c++
void my_function(std::ostream& out) {
  my_function(Fencer<execution_space>("my_function").space, out); 
}
```
can also be written as
```c++
void my_function(std::ostream& out)
{
  Fencer<execution_space> fencer("msg");
  my_function(fencer.space, out);
}
```
(looks like a *scope guarded* thing)

Reply:
Take the discussion to its own issue if you want to gauge for appetite for it.
I am unconvinced.

Reply:
FWIF if someone throws an exception (ugh) in between, at least we will fence properly with this class.

Reply:
Discussion moved to #6716.